os/coreos-layering.md

```diff
@@ -0,0 +1,99 @@
+
+# Existing work
+
+Originally, https://github.com/coreos/fedora-coreos-tracker/issues/812 tracked native support for "encapsulating" ostree commits in containers.
+
+Then, it was realized that when shipping the OS as a container image, it feels natural to support users deriving from it.  The bulk of this is really "ostree native container" integration glue, and is happening in https://github.com/ostreedev/ostree-rs-ext
+
+rpm-ostree vendors the ostree-rs-ext code and also be extended to support the same interfaces as implemented by the "base" ostree-rs-ext code.
+
+Specifically as of today, this functionality is exposed in:
+
+- `rpm-ostree ex-container`
+- `rpm-ostree rebase --experimental $containerref`
+
+# Rationale
+
+Since the creation of Container Linux (CoreOS) as well as Atomic Host, and continuing into the Fedora/RHEL CoreOS days, we have faced a constant tension around what we ship in the host system.
+
+[This issue](https://github.com/coreos/fedora-coreos-tracker/issues/401) encapsulates much prior discussion.
+
+For users who are happy with Fedora CoreOS today, not much will change.
+
+For those who e.g. want to install custom agents or nontrivial amounts of code (such as kubelet), this "CoreOS layering" will be a powerful new mechanism to ship the code they need.
+
+# Example via Dockerfile
+
+[fcos-derivation-example](https://github.com/cgwalters/fcos-derivation-example) contains an example `Dockerfile` that builds a Go binary and injects it along with a corresponding systemd unit as a layer, building on top of Fedora CoreOS.
+
+For ease of reference, a copy of the above is inline here:
+
+```dockerfile
+# Build a small Go program using a builder image
+FROM registry.access.redhat.com/ubi8/ubi:latest as builder
+WORKDIR /build
+COPY . .
+RUN yum -y install go-toolset
+RUN go build hello-world.go
+
+# In the future, this would be e.g. quay.io/coreos/fedora:stable
+FROM quay.io/cgwalters/fcos-dev
+# Inject it into Fedora CoreOS
+COPY --from=builder /build/hello-world /usr/bin
+# And add our unit file
+ADD hello-world.service /etc/systemd/system/hello-world.service
+# Also add strace; we don't yet support `yum install` but we can
+# with some work in rpm-ostree!
+RUN rpm -Uvh https://kojipkgs.fedoraproject.org//packages/strace/5.14/1.fc34/x86_64/strace-5.14-1.fc34.x86_64.rpm
+```
+
+# Derivation versus Ignition/butane
+
+This proposal does not replace Ignition.  Ignition will still play at least two key roles:
+
+- Setting up partitions and storage, e.g. LUKS is something configured via Ignition provided on boot.
+- Machine/node specific configuration, in particular bootstrap configuration: e.g. static IP addresses that are necessary to fetch container images at all.
+
+# Butane as a declarative input format for layering
+
+We must support `Dockerfile`, because it's the lowest common denominator for the container ecosystem, and is accepted as input for many tools.
```

Comment:
To be clear, I'm OK with this too, and I know we can always tweak things going forward (though it's always harder to add restrictions than it is to remove them). Just wanted to point out what IMO are some non-negligible risks of this approach.

Reply:
> we have this conscious decision to separate out pure package layering and base package overrides. This distinction does not exist here, 

That's not quite true, because we will still have the original base image.  Reverting back to the "golden FCOS base" will also be a just an `rpm-ostree rebase` away.

Additionally, because we will have `yum|dnf -> rpm-ostree` in this image (xref https://coreos.github.io/rpm-ostree/cliwrap/) we can actually still impose the `install` vs `override` semantic.

Further, tooling can perform filesystem and rpmdb diffing between layers - we have all that code today.

Reply:
> And they each use completely different languages. 

Also, today one can use any tool to generate Ignition, which definitely happens (e.g. `openshift-install` generates Ignition via some custom Terraform stuff) and we support that.

More broadly I think we are in a position of needing to support "low level" as well as arbitrary configuration mechanisms, but we also try to have opinions on higher level tooling.

So here `Dockerfile` is that lowest common denominator for arbitrary, low level (kinda) mechanism - but we can streamline higher level workflows inside of that.

Reply:
> > we have this conscious decision to separate out pure package layering and base package overrides. This distinction does not exist here,
> 
> That's not quite true, because we will still have the original base image. Reverting back to the "golden FCOS base" will also be a just an `rpm-ostree rebase` away.

Right, but the power a user wields in their `Dockerfile` sees it all as just one filesystem. E.g. nothing stops them from doing `COPY my-libc.so /usr/lib64/libc.so` and the build system will be happy to build that. Whereas if it's Butane based, we can immediately tell they're modifying base content and fail the build. So the failure happens at image build time instead of deploy and reboot time. Of course, we do want to provide flexibility to modify base content, but IMO it should be explicitly opted in.

Combining this with https://github.com/ostreedev/ostree-rs-ext/issues/159, maybe we could have this `ostree container finalize` step take that switch? E.g. `ostree container finalize --allow-overrides`.

> Additionally, because we will have `yum|dnf -> rpm-ostree` in this image (xref [coreos.github.io/rpm-ostree/cliwrap](https://coreos.github.io/rpm-ostree/cliwrap/)) we can actually still impose the `install` vs `override` semantic.

Hmm, can you expand on the UX you're thinking of here? `dnf install` will automatically upgrade an already installed package so we'd have to require some added switch (related: https://github.com/coreos/rpm-ostree/pull/2844#discussion_r647703671).

Reply:
> Hmm, can you expand on the UX you're thinking of here? dnf install will automatically upgrade an already installed package so we'd have to require some added switch (related: coreos/rpm-ostree#2844 (comment)).

Well, the cliwrap status quo today errors actually when someone types `dnf install`.  I know this is a perpetually confusing thing but so far cliwrap does not involve necessarily invoking traditional `yum|dnf` logic.  It opens the door to that of course.

So specifically if e.g. someone times `RUN yum install usbguard` in their `Dockerfile`, we would error out, and they'd have to do `RUN rpm-ostree install usbguard` which would run through all of the same logic that exists today which would only layer, not upgrade dependencies.  (There's...a *lot* implied in this whole "rpm-ostree in container builds" flow, really worth splitting out to a separate issue; will look at that)

Reply:
The real status quo *right now* to be clear is that `dnf install usbguard` will give "command not found" and `rpm-ostree install usbguard` will error out because you're in a container.

Reply:
https://github.com/coreos/rpm-ostree/issues/3227

Reply:
```
> podman run --rm -ti quay.io/cgwalters/fcos bash
bash-5.1# dnf install usbguard
bash: dnf: command not found
bash-5.1# rpm-ostree install usbguard
error: This system was not booted via libostree; found $container=podman environment variable.
rpm-ostree is designed to manage host systems, not containers.

bash-5.1# 
```

Reply:
> So specifically if e.g. someone times `RUN yum install usbguard` in their `Dockerfile`, we would error out, and they'd have to do `RUN rpm-ostree install usbguard` which would run through all of the same logic that exists today which would only layer, not upgrade dependencies.

Ack, I like that.

What I'm trying to make sure here is that OCP customers who will be interacting with this feature are still subject to basic rules by default so that our QA efforts aren't trivially invalidated. I.e. I don't want them to be able to change `/usr` base content just as easily as they can `/etc` content. If the only way OCP customers interact with this is through Butane or MCs for example, then I think that's fine. Otherwise, I think we should make them opt into modifying base content explicitly (and maybe e.g. we only support them doing this if they have a support exception).

Taking this feature in isolation (e.g. for FCOS users), typing `rpm-ostree rebase $my_pullspec` already implies some risk acceptance (though there's still value in keeping guardrails there too of course). But in the OCP case, the MCO will be what's running that command.